README
key-combo.el
========

[![Build Status](https://secure.travis-ci.org/uk-ar/key-combo.png)](http://travis-ci.org/uk-ar/key-combo)


CHANGELOG.md
# Changelog

## 1.6 2015/03/24 15:24:37

* eldoc and company-mode work correctly with key-combo commands
* Partial fix for multiple-cursors
* Several IME fixes and improvements
* Support vector format in definition of key-combos
* Add global-key-combo-mode-map to support global keys
* Enable the use of SPC key
* Add key-combo-disable-modes to disable in some modes

## 1.5.1 2012/06/06 21:36:28
* Bug fix which use flex-autopair by mistake.

## 1.5 2012/04/20 22:24:26
* Bug fix when just after string.
* Add !== for js and php's not triple-equal by tomykaira.
* Change some default settings.

## 1.4.1 2012/04/04 21:05:48
* Bug fix for first key in c-mode and other modes.

## 1.4 2012/04/03 20:15:21
* Regard first key as key-combo-execute-original when first key is not assigned
* Auto indent when inserting string have new line

## 1.3 2012/03/13 22:00:23
* Make works well for other elisp which use post command hook

## 1.2 2012/02/10 22:15:52
* Add support to use SKK. Bug reported by ballforest
* Bug fix for html mode.

## 1.1 2012/02/08 21:56:27
* Add key-combo-define-local function to set key for local keymap.
* Add a lot of default setting in pogin's blog.

## 1.0 2012/01/31 22:03:50
* Change clean-up function to use undo

## 0.7 2012/01/17 21:25:10
* Insert white space dwim

## 0.6 2012/01/16 21:17:01
* Allow cleanup function as nil
* Add key-combo-return function,
which can move to point of command beginning.
* Allow meta key for key-combo key.
* Save undo history when self-insert-command.

## 0.5 2012/01/13 23:02:39
* Support function as key-combo command

## 0.4
* Map key to minor mode to toggle enable and disable.

## 0.3
* Not to cleanup when 1 sequence key
* Bugfix by tomykaira
* Refactoring
* Add test cases

## 0.2
* First release

## 0.1
* Initial revision



test/run-test.el
;; Usage:
;;
;;   emacs -Q -l tests/run-test.el           # interactive mode
;;   emacs -batch -Q -l tests/run-test.el    # batch mode


;; Utils
(defun key-combo-test-join-path (path &rest rest)
  "Join a list of PATHS with appropriate separator (such as /).

\(fn &rest paths)"
  (if rest
      (concat (file-name-as-directory path) (apply 'key-combo-test-join-path rest))
    path))

(defvar key-combo-test-dir (file-name-directory load-file-name))
(defvar key-combo-root-dir (concat key-combo-test-dir ".."))


;; Setup `load-path'
(mapc (lambda (p) (add-to-list 'load-path p))
      (list key-combo-test-dir
            key-combo-root-dir))


;; Use ERT from github when this Emacs does not have it
(unless (locate-library "ert")
  (add-to-list
   'load-path
   (key-combo-test-join-path key-combo-root-dir "lib" "ert" "lisp" "emacs-lisp")))

;; Use el-spec
(add-to-list
 'load-path
 (key-combo-test-join-path key-combo-root-dir "lib" "el-spec"))

;; el-spy
(add-to-list
 'load-path
 (key-combo-test-join-path key-combo-root-dir "lib" "el-spy"))

;; Load tests
(load "key-combo-test")


;; Run tests
(if noninteractive
    (ert-run-tests-batch-and-exit)
  (ert t))


test/run-tests.el
;; Usage:
;;
;;   emacs -Q -l tests/run-test.el           # interactive mode
;;   emacs -batch -Q -l tests/run-test.el    # batch mode


;; Utils
(defun key-combo-test-join-path (path &rest rest)
  "Join a list of PATHS with appropriate separator (such as /).
\(fn &rest paths)"
  (if rest
      (concat (file-name-as-directory path) (apply 'key-combo-test-join-path rest))
    path))

(defvar key-combo-test-dir (file-name-directory load-file-name))
(defvar key-combo-root-dir (concat key-combo-test-dir ".."))

;; Setup `load-path'
(mapc (lambda (p) (add-to-list 'load-path p))
      (list key-combo-test-dir
            key-combo-root-dir))

;; Load tests
(load "key-combo-tests")
(ert-run-tests-batch-and-exit t)


old/key-combo-test.el
(require 'ert)
(require 'el-spec)

(require 'key-combo)
(key-combo-load-default)

(defun test1()
  (interactive)
  (message "test1")
  )

(defun test2()
  (interactive)
  (message "test2")
  )

(defun key-combo-test-helper-execute (cmd)
  (key-combo-mode 1)
  (execute-kbd-macro (key-combo-read-kbd-macro cmd))
  (substring-no-properties (buffer-string)))

(defun key-combo-test-helper-define-lookup (cmd)
  (key-combo-define-global ">>" cmd)
  (key-combo-key-binding ">>"))

(defun key-combo-test-helper-binding-execute (cmd)
  (key-combo-command-execute (key-combo-key-binding cmd))
  (substring-no-properties (buffer-string)))

(dont-compile
  (when (fboundp 'describe)
    (describe ("key-combo in temp-buffer" :vars ((mode)))
      (shared-examples "check pre-command-hook"
        (it ()
          (key-combo-mode 1)
          (should (memq 'key-combo-pre-command-function pre-command-hook)))
        (it ()
          (key-combo-mode -1)
          (should-not (memq 'key-combo-pre-command-function pre-command-hook))))
      (shared-examples "C-a"
        (before
          (insert "B\n IP")
          (key-combo-mode 1))
        (it ()
          (should (key-combo-key-binding (kbd "C-a C-a"))))
        ;; (it ()
        ;;   (key-combo-mode -1)
        ;;   (should-not (key-combo-key-binding (kbd "C-a C-a"))))
        (it ()
          (key-combo-test-helper-execute "C-a")
          (should (equal (char-to-string (following-char)) "I")))
        (it ()
          (key-combo-test-helper-execute "C-a C-a")
          (should (equal (char-to-string (following-char)) " ")))
        (it ()
          (key-combo-test-helper-execute "C-a C-a C-a")
          (should (equal (char-to-string (following-char)) "B")))
        ;; fail in temp buffer?
        ;; (it (:vars ((cmd "C-a C-a C-a C-a")))
        ;;   (backward-char)
        ;;   (should (equal (char-to-string (following-char)) "P")))
        )

      (around
        (setq key-combo-command-keys nil)
        (with-temp-buffer
          (switch-to-buffer (current-buffer))
          (let ((key-combo-mode-map
                 (copy-keymap key-combo-mode-map))
                (global-map-org (current-global-map))
                (global-map (copy-keymap (current-global-map))))
            (unwind-protect
                (progn
                  (use-global-map global-map)
                  (funcall el-spec:example))
              (use-global-map global-map-org)))))

      (it ()
        (key-combo-test-helper-execute "C-a C-a")
        (should (equal (char-to-string (following-char)) " "))
        (should (eq real-last-command 'move-beginning-of-line))
        )
      (it ()
        (key-combo-test-helper-execute "C-a C-a C-a")
        (should (equal (char-to-string (following-char)) "B"))
        (should (eq real-last-command 'beginning-of-buffer))
        )
      (it ()
        (key-combo-test-helper-execute "C-a C-p"))
      ;; fail in temp buffer?
      ;; (it (:vars ((cmd "C-a C-a C-a C-a")))
      ;;   (backward-char)
      ;;   (should (equal (char-to-string (following-char)) "P")))

      (context "in default-mode"
        (context "execute"
          (it ()
            (should (string= (key-combo-test-helper-execute ">") ">")))
          (it ()
            (should (string= (key-combo-test-helper-execute "=") "="))))
        (context "no execute"
          (it ()
            (key-combo-command-execute (lambda () (insert "a")))
            (should (string= (buffer-string) "a")))
          (it ()
            (should-error (key-combo-command-execute 'wrong-command)))
          (it ()
            (let ((last-command-event ?b))
              (key-combo-command-execute 'self-insert-command))
            (should (string= (buffer-string) "b")))
          (it ()
            (key-combo-command-execute (kc-get-command "a"))
            (should (string= (buffer-string) "a")))
          (it ()
            (key-combo-command-execute (kc-get-command "a`!!'a"))
            (should (string= (buffer-string) "aa"))
            (should (eq (point) 2)))
          (it ()
            (buffer-enable-undo)
            (key-combo-mode 1)
            (defmock test1 () (interactive) (insert "test1"))
            (defmock test2 () (interactive) (insert "test2"))
            (defmock test3 () (interactive) (insert "test3"))
            ;; (funcall el-spec:example) ;raise error?
            )
          (it ()
            (define-prefix-command 'test-map)
            (global-set-key (kbd "M-s") 'test-map)
            (global-set-key (kbd "M-s z") 'test3)

            (should (string= (key-combo-test-helper-execute "M-s z")
                             "test3"))
            (should (eq (el-spy:called-count 'test1) 0))
            (should (eq (el-spy:called-count 'test2) 0))
            (should (eq (el-spy:called-count 'test3) 1))
            )
          (it ()
	      (define-prefix-command 'test-map)
	      (global-set-key (kbd "M-s") 'test-map)
	      (define-key test-map (kbd "z")
		'test3)
	      (key-combo-define-global (kbd "M-s") 'test1)
	      (key-combo-define-global (kbd "M-s a") 'test2)
	      (it ()
		  (define-prefix-command 'test-map)
		  (global-set-key (kbd "M-s") 'test-map)
		  (global-set-key (kbd "M-s z") 'test3)

		  (define-key global-map
		    (kc-make-key-vector (kbd "M-s")) 'test1)
		  (define-key global-map
		    (kc-make-key-vector (kbd "M-s a")) 'test2)
		  (should (string= (key-combo-test-helper-execute "M-s a")
				   "test2"))
		  (should (eq (el-spy:called-count 'test1) 1))
		  (should (eq (el-spy:called-count 'test2) 1))
		  (should (eq (el-spy:called-count 'test3) 0))
		  )
	      (it ()
		  (define-prefix-command 'test-map)
		  (global-set-key (kbd "M-s") 'test-map)
		  (define-key test-map (kbd "z")
		    'test3)
		  (define-key global-map
		    (kc-make-key-vector (kbd "M-s"))
		    'test1)
		  (define-key global-map
		    (kc-make-key-vector (kbd "M-s a"))
		    'test2)
		  (should (string= (key-combo-test-helper-execute "M-s")
				   "test1"))
		  (should (eq (el-spy:called-count 'test1) 1))
		  (should (eq (el-spy:called-count 'test2) 0))
		  (should (eq (el-spy:called-count 'test3) 0))
		  )
	      (it ()
		  (define-prefix-command 'test-map)
		  (global-set-key (kbd "M-s") 'test-map)
		  (define-key test-map (kbd "z")
		    'test3)
		  (define-key global-map
		    (kc-make-key-vector (kbd "M-s"))
		    'test1)
		  (define-key global-map
		    (kc-make-key-vector (kbd "M-s a"))
		    'test2)
		  (should (string= (key-combo-test-helper-execute "M-s z")
				   "test3"))
		  (should (eq (el-spy:called-count 'test1) 1))
		  (should (eq (el-spy:called-count 'test2) 0))
		  (should (eq (el-spy:called-count 'test3) 1))
		  )
	      (should (string= (key-combo-test-helper-execute "M-s a")
			       "test2"))
	      (should (eq (el-spy:called-count 'test1) 1))
	      (should (eq (el-spy:called-count 'test2) 1))
	      (should (eq (el-spy:called-count 'test3) 0))
	      )
          (it ("multiple prefix")
            ;; "a M-s"
            )
          (it ("multiple prefix2")
            ;; "M-s M-s"
            )
          (it ("define key")
            ;; "M-s M-s"
            )
          )
        (it ()
          (should-error
           (with-mock
             (mock (test1 *) :times 1)
             (key-combo-define-global (kbd "M-C-d") '(test1 test2)))))
        (it ()
          ;; no error
          (with-el-spy
           (defmock test1 () (interactive))
           (defmock test2 () (interactive))
           (key-combo-mode 1)
           (key-combo-define-global (kbd "M-C-d") 'test1)
           (execute-kbd-macro (kbd "M-C-d"))
           ;; (should (eq (el-spy:called-count 'test1) 1))
           ))
        (it ()
          ;; no error
          (with-el-spy
           (defmock test1 () (interactive))
           (defmock test2 () (interactive))
           (key-combo-mode 1)
           (key-combo-define-global (kbd "M-C-d") '(test1 test2))
           (execute-kbd-macro (kbd "M-C-d"))
           ;; (should (eq (el-spy:called-count 'test1) 1))
           ))
        (it ()
          ;; no error
          (with-el-spy
           (defmock test1 () (interactive))
           (defmock test2 () (interactive))
           (key-combo-mode 1)
           (key-combo-define-global (kbd "M-C-d") '(test1 test2))
           (execute-kbd-macro (kbd "M-C-d M-C-d"))
           (should (eq (el-spy:called-count 'test1) 1))
           (should (eq (el-spy:called-count 'test2) 1))
           ))
        (it ()
          (insert ";\n")
          (should-not (kc-in-string-or-comment-p)))
        (context "isearch-mode"
          (it ()
            (insert "=")
            (should (string= (buffer-string) "="))
            (should (eq (point) 2)))
          (it ()
            (insert "=");; not to raise error from isearch-search
            (isearch-mode nil);; backward search
            (execute-kbd-macro "=")
            (should (string= (buffer-string) "="))
            (should (eq (point) 1))))
        (context "execute only"
          (it ()
            (should (string= (key-combo-test-helper-execute "=") "= ")))
          (it ()
            (should (string= (key-combo-test-helper-execute "==") "eq ")))
          (it ()
            (should (string= (key-combo-test-helper-execute ",") ",")))
          (it ()
            (should (string= (key-combo-test-helper-execute ",,") ",,")))
          (it ()
            (should (string= (key-combo-test-helper-execute ".") ".")))
          (it ()
            (should (string= (key-combo-test-helper-execute ". SPC") " . ")))
          (it ()
            (should (string= (key-combo-test-helper-execute ";") ";; ")))
          (it ()
            (should (string= (key-combo-test-helper-execute ";.") ";; .")))
          (it ()
            (should (string= (key-combo-test-helper-execute ";,") ";; ,")))
          (it ()
            (insert ";")
            (should (string= (key-combo-test-helper-execute "=") ";=")))
          )
        (context "with mock"
          (when (require 'el-mock nil t)
            (it ()
              (should-error
               (with-mock
                 (mock (test1 *) :times 1)
                 (key-combo-define-global (kbd "M-C-d") '(test1 test2)))))
            (it ()
              ;; no error
              (with-mock
                (mock (test1 *) :times 1)
                (key-combo-define-global (kbd "M-C-d") '(test1 test2))
                (execute-kbd-macro (kbd "M-C-d"))))
            (it ()
              ;; no error
              (with-mock
                (mock (test1 *) :times 1)
                (mock (test2 *) :times 1)
                (key-combo-define-global (kbd "M-C-d") '(test1 test2))
                (execute-kbd-macro (kbd "M-C-d M-C-d"))))
            (it ()
              ;; no error
              (with-mock
                (mock (define-key * * *) :times 1)
                (key-combo-define-local "a" "a")))
            (it ()
              ;; no error
              (with-mock
                (mock (define-key * * *) :times 1)
                (key-combo-define-local "a" '("a"))))
            (it ()
              ;; no error
              (with-mock
                (mock (define-key * * *) :times 3);; 1 for recursive call?
                (key-combo-define-local "a" '("a" "b"))))
            (it ()
              ;; no error
              (with-mock
                (mock (lookup-key * *) => t :times 2)
                (mock (define-key * * *) :times 2);; 1 for recursive call?
                (key-combo-define-local "a" '("a" "b")))))
          )
        (context "in skk-mode"
          (when (require 'skk-autoloads nil t)
            (before
              (skk-mode 1)
              (setq this-command 'skk-insert)
              (insert ";")
              )
            (it ()
              (should (string= (key-combo-test-helper-execute ",") ";、")))
            (it ()
              (should (string= (key-combo-test-helper-execute ".") ";。")))
            )
          )
        (it ()
          (insert "init:")
          (should (string= (key-combo-test-helper-execute "=") "init:= "))
          (undo -1)
          (should (string= (buffer-string) "init:="))
          (undo-more 1)
          (should (string= (buffer-string) ""))
          )
        (include-examples "C-a")
        (include-examples "check pre-command-hook"))
      (context "in ruby"
        (before
          (key-combo-mode 1)
          (ruby-mode)
          (when (boundp 'auto-complete-mode)
            (auto-complete-mode -1)))
        ;; bug?for auto-complete completion
        (it ()
          (should (string= (key-combo-test-helper-execute ".") ".")))
        (it ()
          (should (string= (key-combo-test-helper-execute ". SPC") " . ")))
        (it ()
          (should (string= (key-combo-test-helper-execute ";") ";; ")))
        (it ()
          (should (string= (key-combo-test-helper-execute ";.") ";; .")))
        (it ()
          (should (string= (key-combo-test-helper-execute ";,") ";; ,")))
        (it ()
          (should (string= (key-combo-test-helper-execute "||=") " ||= "))))
      (context "in c-mode"
        (before
          ;; (key-combo-mode 1)
          (c-mode))
        (context "execute+"
          (it ()
            (should (string= (key-combo-test-helper-execute "+") " + ")))
          (it ()
            (should (string= (key-combo-test-helper-execute "++") "++")))
          ;; (it ()
          ;;   (should (string= (key-description "+") "+")))
          (it ()
            (should (equal (listify-key-sequence "+") '(43))))
          (it ()
            (should (string= (key-description '(?+)) "+")))
          (it ()
            (should (equal (kc-make-key-vector '(?+))
                           ;;(vector 'key-combo (intern (key-description )))
                           [key-combo _+])))
          (it ("a")
            (should (not (null (key-binding
                                (kc-make-key-vector '(?+))
                                )))))
          (it ("c")
            (should (not (null (lookup-key
                                (current-local-map)
                                (kc-make-key-vector '(?+))
                                )))))
          (it ("b")
            (should (not (equal (key-binding
                                 (kc-make-key-vector '(?+)))
                                'key-combo-execute-original))))
          (it ()
            (should (not (null (kc-get-command "+")))))
          (it ()
            (should (not (equal (kc-get-command "+")
                                'key-combo-execute-original))))
          (it ("d")
            (key-combo-define-local "a" nil)
            ;; (key-combo-key-binding "a")
            ;; (key-binding (vector 'key-combo (intern (key-description "a"))))
            ;; accept-default bug?
            (should (eq (lookup-key (current-local-map)
                                    (kc-make-key-vector '(?a)))
                        nil))
            (key-combo-define-local "a" "a")
            (should (not (equal (lookup-key (current-local-map)
                                            (kc-make-key-vector '(?a)))
                                nil)))
            (key-combo-define-local "a" nil)
            )
          )
        (context "undo"
          (before
            (buffer-enable-undo))
          (it ()
            (should (string= (key-combo-test-helper-execute "=") " = ")))
          (it ()
            (should (string= (key-combo-test-helper-execute "=") " = "))
            (undo)
            (should (string= (buffer-string) "="))
            )
          (it ()
            (should (string= (key-combo-test-helper-execute ",") ";、")))
          (it ()
            (should (string= (key-combo-test-helper-execute ".") ";。")))
          )
        )
      (context ("insert & move & execute" :vars (pos pre-string))
        (it ()
          (insert "\"")
          (should (string= (key-combo-test-helper-execute "=") "\"=")))
        (it ()
          (insert ";")
          (should (string= (key-combo-test-helper-execute "=") ";=")))
        (it ()
          (insert ";")
          (should (string= (key-combo-test-helper-execute ",") ";,")))
        (it ()
          (insert ";\n")
          (should (string= (key-combo-test-helper-execute ";") ";\n;; ")))
        (it ()
          (insert ";")
          (should (string= (key-combo-test-helper-execute ".") ";.")))
        (it ()
          (insert "\"\"\n")
          (goto-char 3)
          (should (string= (key-combo-test-helper-execute ".") "\"\".\n")))
        (it ()
          (insert "\"\"a")
          (goto-char 3)
          (should (string= (key-combo-test-helper-execute ".") "\"\".a")))
        (it ()
          (insert "\"\"")
          (goto-char 3)
          (should (string= (key-combo-test-helper-execute ".") "\"\".")))
        (it ()
          (insert "\"\"")
          (goto-char 2)
          (should (string= (key-combo-test-helper-execute ".") "\".\"")))
        (it ()
          (insert "a\"\"")
          (goto-char 2)
          (should (string= (key-combo-test-helper-execute ".") "a.\"\"")))
        )
      (include-examples "C-a")
      (include-examples "check post-command-hook"))
    (context "in ruby"
      (before
        (buffer-enable-undo)
        (key-combo-mode 1)
        (ruby-mode)
        (when (boundp 'auto-complete-mode)
          (auto-complete-mode -1)))
      ;; bug?for auto-complete completion
      (it ()
        (should (string= (key-combo-test-helper-execute ".") ".")))
      (it ()
        (should (string= (key-combo-test-helper-execute "..") "..")))
      (it ()
        (should (string= (key-combo-test-helper-execute "...") "...")))
      (it ()
        (should (string= (key-combo-test-helper-execute "!~") " !~ ")))
      (it ()
        (should (string= (key-combo-test-helper-execute "**") "**")))
      (it ()
        (should (string= (key-combo-test-helper-execute "||=") " ||= "))))
    (context "in c-mode"
      (before
        (buffer-enable-undo)
        ;; (key-combo-mode 1)
        (c-mode))
      (context "execute+"
        (it ()
          (should (string= (key-combo-test-helper-execute "+") " + ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "++") "++")))
        ;; (it ()
        ;;   (should (string= (key-description "+") "+")))
        (it ()
          (should (equal (listify-key-sequence "+") '(43))))
        (it ()
          (should (string= (key-description '(?+)) "+")))
        (it ()
          (should (equal (kc-make-key-vector '(?+))
                         ;;(vector 'key-combo (intern (key-description )))
                         [key-combo _+])))
        (it ("a")
          (should (not (null (key-binding
                              (kc-make-key-vector '(?+))
                              )))))
        (it ("c")
          (should (not (null (lookup-key
                              (current-local-map)
                              (kc-make-key-vector '(?+))
                              )))))
        (it ("b")
          (should (not (equal (key-binding
                               (kc-make-key-vector '(?+)))
                              'key-combo-execute-original))))
        (it ()
          (should (not (null (kc-get-command "+")))))
        (it ()
          (should (not (equal (kc-get-command "+")
                              'key-combo-execute-original))))
        (it ("d")
          (key-combo-define-local "a" nil)
          ;; (key-combo-key-binding "a")
          ;; (key-binding (vector 'key-combo (intern (key-description "a"))))
          ;; accept-default bug?
          (should (eq (lookup-key (current-local-map)
                                  (kc-make-key-vector '(?a)))
                      nil))
          (key-combo-define-local "a" "a")
          (should (not (equal (lookup-key (current-local-map)
                                          (kc-make-key-vector '(?a)))
                              nil)))
          (key-combo-define-local "a" nil)
          )
        (context "funcall"
          ;; (before
          ;;   (key-combo-command-execute (key-combo-key-binding lookup-cmd)))
          (it ()
            (should (string=
                     (key-combo-test-helper-binding-execute "=") " = ")))
          (it ()
            (should (string=
                     (key-combo-test-helper-binding-execute "==") " == ")))
          (it ()
            (should (string=
                     (key-combo-test-helper-binding-execute [?=]) " = ")))
          (it ()
            (should (string=
                     (key-combo-test-helper-binding-execute [?= ?=]) " == ")))
          (it ()
            (should (string=
                     (key-combo-test-helper-binding-execute [?= ?>]) " => ")))
          (it ()
            (should (string=
                     (key-combo-test-helper-binding-execute [?= ?= ?=])
                     " === ")))
          ;; (it ()
          ;;   (funcall (key-combo-key-binding [?= ?= ?= ?=]))
          ;;   (should (string= (buffer-string) " ==== ")))
          (it ()
            (key-combo-define-global (kbd "C-M-h") " == ")
            (key-combo-command-execute (key-combo-key-binding (kbd "C-M-h")))
            (should (equal (buffer-string) " == ")))
          (it ()
            (should-not
             (equal
              (key-combo-lookup-key (current-global-map) (kbd "C-M-h"))
              " == ")))
          ;; (it ()
          ;;   (key-combo-define-global (kbd "C-M-h C-M-h") " === ")
          ;;   (execute-kbd-macro (kbd "C-M-h C-M-h"))
          ;;   (should (string= (buffer-string) " === "))
          ;;   )
          (it ()
            (key-combo-define-global (kbd "C-M-h C-M-h") " === ")
            (key-combo-command-execute
             (key-combo-key-binding (kbd "C-M-h C-M-h")))
            (should (string= (buffer-string) " === "))
            )
          (it ()
            (should-not (key-combo-key-binding [?= ?= ?= ?=])))
          (it ()
            (insert "a  ")
            (should (string= (key-combo-test-helper-execute "=") "a  = "))
            (should (string= (buffer-string) "a  = ")))
          )
        (it ()
          (should (string= (key-combo-test-helper-execute "= C-x u") "=")))
        (it ()
          (should (string= (key-combo-test-helper-execute "== C-x u") " = ")))
        )
      (context "execute"
        (it ()
          (should (string= (key-combo-test-helper-execute "=") " = ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "=*") " =* ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "==") " == ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "===") " === ")))
        (it "loop"
          (should (string= (key-combo-test-helper-execute "====") " = ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "=>=") " => = ")))
        ;; (it ()
        ;; (should (string= (key-combo-test-helper-execute "==!") " ==! ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "=>") " => ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "/") "/")))
        (it ()
          (should (string= (key-combo-test-helper-execute "/ SPC") " / ")))
        (it ()
          (should (string= (key-combo-test-helper-execute "*") "*")))
        (it ()
          (should (string= (key-combo-test-helper-execute "**") "**")))
        (it ()
          (should (string= (key-combo-test-helper-execute "->") "->")))
        (it ()
          (should (string= (key-combo-test-helper-execute ".") ".")))
        ;; todo check position
        (it ()
          (should (string= (key-combo-test-helper-execute "/* RET")
                           "/*\n  \n */")))
        ;; todo depend on indent width
        ;; (it ()
        ;; (should (string= (key-combo-test-helper-execute "{ RET") "{\n  \n}"))
        )
      (context "funcall"
        ;; (before
        ;;   (key-combo-command-execute (key-combo-key-binding lookup-cmd)))
        (it ()
          (should (string=
                   (key-combo-test-helper-execute "=") " = ")))
        (it ()
          (should (string=
                   (key-combo-test-helper-execute "==") " == ")))
        (it ()
          (should (string=
                   (key-combo-test-helper-execute "=>") " => ")))
        (it ()
          (should (string=
                   (key-combo-test-helper-execute "===")
                   " === ")))
        ;; (it ()
        ;;   (funcall (key-combo-key-binding [?= ?= ?= ?=]))
        ;;   (should (string= (buffer-string) " ==== ")))
        ;; (it ()
        ;;   (key-combo-define-global (kbd "C-M-h") " == ")
        ;;   (key-combo-command-execute (key-combo-key-binding (kbd "C-M-h")))
        ;;   (should (equal (buffer-string) " == ")))
        (it ()
          (should-not
           (equal
            (key-combo-lookup-key (current-global-map) (kbd "C-M-h"))
            " == ")))
        ;; (it ()
        ;;   (key-combo-define-global (kbd "C-M-h C-M-h") " === ")
        ;;   (execute-kbd-macro (kbd "C-M-h C-M-h"))
        ;;   (should (string= (buffer-string) " === "))
        ;;   )
        ;; (it ()
        ;;   (key-combo-define-global (kbd "C-M-h C-M-h") " === ")
        ;;   (key-combo-command-execute
        ;;    (key-combo-key-binding (kbd "C-M-h C-M-h")))
        ;;   (should (string= (buffer-string) " === "))
        ;;   )
        (it ()
          (should-not (key-combo-key-binding [?= ?= ?= ?=])))
        (it ()
          (insert "a  ")
          (should (string= (key-combo-test-helper-execute "=") "a  = "))
          (should (string= (buffer-string) "a  = ")))
        )
      )
    ))


key-combo-config.el
(require 'key-combo)


;; (defcustom key-combo-loop 'only
;;   "Loop mode setting.
;;  'allways: always loop
;;  'only: loop on same key sequence onlyl
;;  'never: don't loop")

;; (defvar key-combo-debug nil)

(defvar key-combo-lisp-mode-hooks
  '(lisp-mode-hook
    emacs-lisp-mode-hook
    lisp-interaction-mode-hook
    inferior-gauche-mode-hook
    scheme-mode-hook))

;; < { [ should use flex-autopair
(defvar key-combo-global-default
  '(;; instead of using (goto-char (point-min))
    ;; use beginning-of-buffer for keydescription
    ("C-a"   . (back-to-indentation move-beginning-of-line
                                    beginning-of-buffer key-combo-return))
    ("C-e"   . (move-end-of-line end-of-buffer key-combo-return))))

(defvar key-combo-lisp-default
  '(("."  . (key-combo-execute-original))
    (". SPC" . " . ")
    ("SPC"  . (key-combo-execute-original))
    ("SPC ." . " . ")
    (","  . (key-combo-execute-original))
    (",@" . " ,@");; for macro
    (";"  . ";; ")
    ;; (";"  . (";; " ";;; " "; ")) ;cannot use because of comment
    (";=" . ";=> ")
    ("="  . ("= " "eq " "equal "))
    (">=" . ">= ")
    ("C-M-x" . (key-combo-execute-original
                (lambda ()
                  (let ((current-prefix-arg '(4)))
                    (call-interactively 'eval-defun)))));; lamda for message
    ("-"  . (key-combo-execute-original));; for symbol name
    ;; ("/" . ("/`!!'/" "/* `!!' */") );;for regexp, comment
    ))

;; (browse-url "http://bojovs.github.com/2012/04/24/ruby-coding-style/")
(defcustom key-combo-common-default
  '((","  . ", ")
    ("="  . (" = " " == " " === " ));;" === " for js
    ("=>" . " => ")
    ("=~" . " =~ ");;for ruby regexp
    ("=*" . " =* ")                     ;for c
    ("+"  . (" + " "++"))
    ("+=" . " += ")
    ("-"  . (" - " "--"))               ;undo when unary operator
    ("-=" . " -= ")
    ("->" . " -> ");; for haskell,coffee script. overwrite in c
    (">"  . (key-combo-execute-original " >> "))
    ;; " > " should be bind in flex-autopair
    (">=" . " >= ")
    (">>=" . " >>= ")
    ("%"  . " % ")
    ("%="  . " %= ")
    ("^"  . " ^ ");; XOR for c
    ("^="  . " ^= ");; for c
    ("!" . key-combo-execute-original)
    ;; NOT for c
    ;; don't use " !" because of ruby symbol
    ;; and unary operator
    ("!="  . " != " ) ;;" !== " for js and php
    ("!==" . " !== ") ;;" !== " for js and php
    ("!~" . " !~ ")   ; for ruby
    ("~" . key-combo-execute-original)
    ;; for unary operator
    ("::" . " :: ") ;; for haskell
    ;; (":" . ":");;for ruby symbol
    ("&"  . (" & " " && "))             ;overwrite in c
    ("&=" . " &= ");; for c
    ("&&=" . " &&= ")                   ; for ruby
    ("*"  . " * " )                     ;overwrite in c
    ("*="  . " *= " )
    ("**"  . "**" )                     ;for power
    ("**=" . " **=" )                     ;for power
    ;; ("?" . "? `!!' :"); ternary operator should be bound in yasnippet?
    ;; ("?=");; for coffeescript?
    ("<" . (key-combo-execute-original " << "))
    ;; " < " should be bound in flex-autopair
    ("<=" . " <= ")
    ;; ("<?" . "<?`!!'?>");; for what?
    ("<<=" . " <<= ");; bit shift for c
    ("<-" . " <- ")
    ("<!" . "<!-- `!!' -->");; for html comment
    ("|"  . (" | " " || "));; bit OR and OR for c
    ;;ToDo: ruby block
    ("|=" . " |= ");; for c
    ("||=" . " ||= ")                   ; for ruby
    ;; ("/" . (" / " "// " "/`!!'/")) ;; devision,comment start or regexp
    ("/" . (key-combo-execute-original))
    ("/ SPC" . " / ")
    ("/=" . " /= ")
    ("*/" . "*/")
    ("/*" . "/* `!!' */")
    ("/* RET" . "/*\n`!!'\n*/");; add *? m-j
    ;; ("/* RET" . "/*\n*`!!'\n*/");; ToDo:change style by valiable
    ("{" . (key-combo-execute-original))
    ("{ RET" . "{\n`!!'\n}"))
  "Default binding which enabled by `key-combo-common-mode-hooks'"
  :group 'key-combo)

(defcustom key-combo-org-default
  '(("C-a" . (org-beginning-of-line
              beginning-of-buffer
              key-combo-return));;back-to-indentation
    ("C-e" . (org-end-of-line
              end-of-buffer
              key-combo-return)))
  "Default binding which enabled by `org-mode-hook'"
  :group 'key-combo)

(defcustom key-combo-pointer-default
  '(("*" . ("*" "**" "***"))
    ("&" . ("&" "&&" "&&&"))
    ("->" . "->"))
  "Default binding for c-mode,c++-mode,objc-mode"
  :group 'key-combo)


;; for algol like language
(defcustom key-combo-common-mode-hooks
  '(c-mode-common-hook;; It's run immediately before the language specific hook.
    php-mode-hook
    ruby-mode-hook
    cperl-mode-hook
    javascript-mode-hook
    js-mode-hook
    js2-mode-hook)
  "Hooks that enable `key-combo-common-default' setting."
  :group 'key-combo)

;;;###autoload
(defun key-combo-load-default ()
  (interactive)
  (global-key-combo-mode t)
  (key-combo-define-kalist (current-global-map)
                            key-combo-global-default)
  (key-combo-define-hook key-combo-common-mode-hooks
                         'key-combo-common-load-default
                         key-combo-common-default)
  (key-combo-define-hook key-combo-lisp-mode-hooks
                         'key-combo-lisp-load-default
                         key-combo-lisp-default)
  (key-combo-define-hook '(c-mode-hook c++-mode-hook)
                         'key-combo-pointer-load-default
                         key-combo-pointer-default)
  (key-combo-define-hook 'objc-mode-hook
                         'key-combo-objc-load-default
                         (append key-combo-pointer-default
                                 '(("@"  . "@\"`!!'\""))))
  (key-combo-define-hook 'org-mode-hook
                         'key-combo-org-load-default
                         key-combo-org-default)
  (key-combo-define-hook '(html-mode-hook
                           css-mode-hook
                           javascript-mode-hook
                           js-mode-hook
                           makefile-mode-hook
                           js2-mode-hook)
                         'key-combo-property-default
                         '((":"  . ": "))))

(provide 'key-combo-config)
